File: java-functional/src/test/java/no/knowit/java_functional/order/OrderServicesTest.java
package no.knowit.java_functional.order;

import static ch.lambdaj.Lambda.*;
import static fj.data.List.*;
import static org.hamcrest.CoreMatchers.*;
import static org.hamcrest.number.OrderingComparisons.*;
import static org.junit.Assert.*;
import static org.junit.matchers.JUnitMatchers.*;

import java.util.Collection;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

import org.joda.time.DateTime;
import org.junit.Ignore;
import org.junit.Test;

import ch.lambdaj.function.matcher.LambdaJMatcher;

import fj.F;
import fj.control.parallel.Strategy;

public class OrderServicesTest implements OrderService {

	/**
	 * Exercise 3: Create orders from train journeys. Use the createOrder
	 * method below to convert individual journeys, to simulate work.
	 */
	@Override
	public Collection<Order> createOrderAlternatives(List<TrainJourney> itineraries) {
		return Collections.emptyList();
	}

	/**
	 * Bonus: Transform train journeys to orders in parallel, to reduce running time.
	 */
	public Collection<Order> createOrdersParallel(List<TrainJourney> itineraries) {
		return createOrderAlternatives(itineraries);
	}

	/**
	 * Method to convert a single train journey to order, simulating
	 * time-consuming work.
	 */
	private Order createOrder(TrainJourney journey) {
		try {
			Thread.sleep(200);
		} catch (InterruptedException e) {
		}
		return new Order(journey, 123.0);
	}

	/**
	 * Plain Java for-loop solution, for reference
	 */
	public Collection<Order> createOrdersLoop(List<TrainJourney> itineraries) {
		List<Order> alternatives = new LinkedList<Order>();
		for (TrainJourney trainJourney : itineraries) {
			alternatives.add(createOrder(trainJourney));
		}
		return alternatives;
	}

	/*
	 * Tests below here
	 */

	@Test
    @Ignore
	public void convert_trainjourney_to_order() {
		List<TrainJourney> itineraries = createItineraries("OSLO S", "BERGEN", new DateTime());

		Collection<Order> orderAlternatives = createOrderAlternatives(itineraries);

		assertThat(orderAlternatives.size(), is(equalTo(5)));
		assertThat("Order missing price", orderAlternatives, everyItem(hasPrice()));
	}

	@Test
    @Ignore
	public void convert_trainjourney_to_order_loop() {
		List<TrainJourney> itineraries = createItineraries("OSLO S", "BERGEN", new DateTime());

		Collection<Order> orderAlternatives = createOrdersLoop(itineraries);

		assertThat(orderAlternatives.size(), is(equalTo(5)));
		assertThat("Order missing price", orderAlternatives, everyItem(hasPrice()));
	}

	@Test
    @Ignore
	public void create_orders_within_timelimit() {
		List<TrainJourney> itineraries = createItineraries("OSLO S", "BERGEN", new DateTime());

		long startTime = System.currentTimeMillis();
		Collection<Order> orderAlternatives = createOrdersParallel(itineraries);
		long endTime = System.currentTimeMillis();
		int duration = (int) (endTime - startTime);

		assertThat(orderAlternatives.size(), is(equalTo(5)));
		assertThat("Order missing price", orderAlternatives, everyItem(hasPrice()));
		assertThat("Create order alternatives too fast", duration, is(greaterThan(200)));
		assertThat("Create order alternatives took too long", duration, is(lessThan(600)));
	}

	private LambdaJMatcher<Order> hasPrice() {
		return having(on(Order.class).getPrice(), greaterThan(0.0));
	}

	private List<TrainJourney> createItineraries(String from, String to, DateTime departureTime) {
		List<TrainJourney> itineraries = new LinkedList<TrainJourney>();
		for (int i = 0; i < 5; i++) {
			itineraries.add(new TrainJourney(from, to, departureTime, departureTime.plusHours(2)));
			departureTime = departureTime.plusHours(1);
		}
		return itineraries;
	}

}

File: scala-functional/src/test/scala/no/knowit/scala_functional/order/OrderServicesTest.scala
package no.knowit.scala_functional.order

import org.hamcrest.CoreMatchers._
import org.hamcrest.number.OrderingComparisons._
import org.joda.time.DateTime
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.junit.Test
import org.junit.Assert._
import org.junit.matchers.JUnitMatchers._
import no.knowit.java_functional.order.Order
import no.knowit.java_functional.order.TrainJourney
import org.junit.Ignore

@RunWith(classOf[JUnit4])
class OrderServicesTest {

  /**
   * Exercise 3: Create orders from train journeys. Use the createOrder
   * method below to convert individual journeys, to simulate work.
   */
  def createOrderAlternatives(itineraries: Seq[TrainJourney]): Seq[Order] = Nil

  /**
   * Bonus: Transform train journeys to orders in parallel, to reduce running time.
   */
  def createOrdersParallel(itineraries: Seq[TrainJourney]): Seq[Order] = Nil

  private def createOrder(journey: TrainJourney): Order = {
    try {
      Thread.sleep(200)
    } catch {
      case e: InterruptedException => // Do nothing
    }
    new Order(journey, 123.0)
  }

  @Test
  @Ignore
  def convert_trainjourney_to_order() {
    val itineraries = createItineraries("OSLO S", "BERGEN", new DateTime())

    val orderAlternatives = createOrderAlternatives(itineraries)

    assertThat(orderAlternatives.size, is(equalTo(5)))
    assertTrue("Order missing price", orderAlternatives forall { _.getPrice > 0.0 })
  }

  @Test
  @Ignore
  def create_orders_within_timelimit() {
    val itineraries = createItineraries("OSLO S", "BERGEN", new DateTime());

    val startTime = System.currentTimeMillis();
    val orderAlternatives = createOrdersParallel(itineraries);
    val endTime = System.currentTimeMillis();
    val duration = endTime - startTime;

    assertThat(orderAlternatives.size, is(equalTo(5)));
    assertTrue("Order missing price", orderAlternatives forall { _.getPrice > 0.0 })
    assertTrue("Create order alternatives too fast", duration >= 200);
    assertTrue("Create order alternatives took too long", duration < 800);
  }

  private def createItineraries(from: String, to: String, departureTime: DateTime): Seq[TrainJourney] =
    for (i <- 1 to 5)
      yield new TrainJourney(from, to, departureTime.plusHours(i), departureTime.plusHours(i + 2))

}